### ISSUE TITLE
foreach-object multi scriptblock example

### ISSUE THREAD

**Issue by jszabo98** [MicrosoftDocs/PowerShell-Docs#4513]
A nice example of passing a scriptblock array to foreach-object -process:

```powershell
Get-ChildItem | ForEach-Object {$sum=0} {$sum++} {$sum}
```

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: e1fdd630-074d-95a8-8da5-ae85b67e7793
* Version Independent ID: 7ba53b91-ed64-cc24-0cf0-a5dc3f3dc9d0
* Content: [ForEach-Object (Microsoft.PowerShell.Core)](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/foreach-object?view=powershell-6)
* Content Source: [reference/6/Microsoft.PowerShell.Core/ForEach-Object.md](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/ForEach-Object.md)
* Product: **powershell**
* Technology: **powershell-cmdlets**
* GitHub Login: @SteveL-MSFT
* Microsoft Alias: **slee**

**Comment by mklement0** [comment]
It's important to clarify that while 3 script blocks passed _positionally_ all _technically_ bind to the
`-Process` parameter, they are effectively treated as if they had been passed to `-Begin`, `-Process`, and `-End`, respectively, as the following example demonstrates:

```powershell
PS> 1..2 | ForEach-Object { 'begin' } { 'process' } { 'end' }
begin
process
process
end
```

That is, the above is (in effect) equivalent to
`1..2 | ForEach-Object -Begin { 'begin' } -Process { 'process' } -End { 'end' }`

---

If you specify _4 or more_ script blocks positionally,  _all interior_ blocks become per-input-object script blocks (though it's unclear when that's ever useful):

```powershell
PS> 1..2 | ForEach-Object { 'begin' } { 'process' } { 'process too' }  { 'end' } 
begin
process
process too
process
process too
end
```

That is, the above is (in effect) equivalent to
`1..2 | ForEach-Object -Begin { 'begin' } -Process { 'process' }, { 'process too' } -End { 'end' }`

**Comment by joeyaiello** [comment]
Looks like `-Process` is special-cased as @mklement0 described, essentially breaking any array of 3 or more script blocks into its own `begin` and `end` blocks:

```powershell
$a = { 'begin' },{ 'process' },{ 'end' }
1..2 | foreach-object -process $a

begin
process
process
end 
```

Two script block arguments treat the first one as a begin block:

```powershell
$c = { 'process' },{ 'process too' }
1..2 | foreach-object -process $c

process
process too
process too
```

And the positionality above works because `-RemainingScripts` has the attribute `ValueFromRemainingArguments`

**Comment by mklement0** [comment]
Another thing worth pointing out about @jszabo98's original example, `Get-ChildItem | ForEach-Object {$sum=0} {$sum++} {$sum}`:

While this is a nice way to perform the initialization of `$sum` as part of the pipeline (and also to output its _value_ via the (effective) `end` block), users should be a aware that the variable is still created in the _caller's_ scope, because that is where all `ForEach-Object` script blocks run.

In other words: `{$sum=0}` may update a preexisting `$sum` variable, and `$sum` will linger after execution of this pipeline.

**Comment by mklement0** [comment]
To add to the above: that 3 script blocks _technically_ bind to `-Process`, only to then in effect be treated as `-Begin`, `-Process` and `-End` respectively, is especially confusing if you use `-Process` _explicitly_, with an _array_ of script blocks:

```powershell
# Despite the use of -Process with an array, the first script block becomes the -Begin
# script block, and the last the -End script block.
1..2 | ForEach-Object -Process { 'one' }, { 'two' }, { 'three' }
one 
two
two
three
```

The reasonable expectation is that all of these are treated as `-Process` blocks - i.e. blocks executed _for each input object_ - yet that is not what happens.

---

In fact, if you do want multiple true (per-input-object) script blocks, you're currently forced to also specify `-Begin` and `-End` explicitly - even if you don't ned them:

```
# For all -Process blocks to be per-input-object blocks, you must also specify -Begin and -End
1..2 | ForEach-Object -Begin $null  -Process { 'one' }, { 'two' }, { 'three' } -End $null

one
two
three
one
two
three
```

A potential use case is wanting to perform conceptually unrelated operations on an input collection while preferring to keep them in separate script blocks for conceptual clarity.

**Comment by sdwheeler** [comment]
@theJasonHelmick Can you help update the docs here?
